pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -858,105 +858,186 @@ func TestPriorityQueue_Pop(t *testing.T) {
-	if _, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(unschedulablePodInfo.Pod)); !exists {
-		t.Errorf("Expected %v to be added to activeQ.", unschedulablePodInfo.Pod.Name)
-	}
-	// Updating a pod that is already in activeQ, should not change it.
-	q.Update(logger, unschedulablePodInfo.Pod, unschedulablePodInfo.Pod)
-	if len(q.unschedulablePods.podInfoMap) != 0 {
-		t.Error("Expected unschedulablePods to be empty.")
-	}
-	if _, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(unschedulablePodInfo.Pod)); !exists {
-		t.Errorf("Expected: %v to be added to activeQ.", unschedulablePodInfo.Pod.Name)
-	}
-	if p, err := q.Pop(); err != nil || p.Pod != highPriNominatedPodInfo.Pod {
-		t.Errorf("Expected: %v after Pop, but got: %v", highPriorityPodInfo.Pod.Name, p.Pod.Name)
-	}
 
-	// Updating a pod that is in backoff queue and it is still backing off
-	// pod will not be moved to active queue, and it will be updated in backoff queue
-	podInfo := q.newQueuedPodInfo(medPriorityPodInfo.Pod)
-	if err := q.podBackoffQ.Add(podInfo); err != nil {
-		t.Errorf("adding pod to backoff queue error: %v", err)
-	}
-	q.Update(logger, podInfo.Pod, podInfo.Pod)
-	rawPodInfo, err := q.podBackoffQ.Pop()
-	podGotFromBackoffQ := rawPodInfo.(*framework.QueuedPodInfo).Pod
-	if err != nil || podGotFromBackoffQ != medPriorityPodInfo.Pod {
-		t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPodInfo.Pod.Name, podGotFromBackoffQ.Name)
-	}
+	tests := []struct {
+		name         string
+		prepareFunc  func(q *PriorityQueue) (oldPod, newPod *v1.Pod)
+		validateFunc func(q *PriorityQueue, expectedPod *v1.Pod)
+	}{
+		{
+			name: "add highPriorityPodInfo to activeQ",
+			prepareFunc: func(q *PriorityQueue) (oldPod, newPod *v1.Pod) {
+				return nil, highPriorityPodInfo.Pod
+			},
+			validateFunc: func(q *PriorityQueue, expectedPod *v1.Pod) {
+				if _, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(expectedPod)); !exists {
+					t.Errorf("Expected %v to be added to activeQ.", expectedPod.Name)
+				}
+				if len(q.nominator.nominatedPods) != 0 {
+					t.Errorf("Expected nomindatePods to be empty: %v", q.nominator)
+				}
+			},
+		},
+		{
+			name: "Update highPriorityPodInfo and add a nominatedNodeName to it",
+			prepareFunc: func(q *PriorityQueue) (oldPod, newPod *v1.Pod) {
+				return highPriorityPodInfo.Pod, highPriNominatedPodInfo.Pod
+			},
+			validateFunc: func(q *PriorityQueue, expectedPod *v1.Pod) {
+				if q.activeQ.Len() != 1 {
+					t.Error("Expected only one item in activeQ.")
+				}
+				if len(q.nominator.nominatedPods) != 1 {
+					t.Errorf("Expected one item in nomindatePods map: %v", q.nominator)
+				}
+			},
+		},
+		{
+			name: "Updating a pod that is already in activeQ, should not change it",
+			prepareFunc: func(q *PriorityQueue) (oldPod, newPod *v1.Pod) {
+				// Updating an unschedulable pod which is not in any of the two queues, should
+				// add the pod to activeQ.
```

Comment:
Can we make a new test case where we can confirm that Update() will just add Pod even if `oldPod` doesn't exist in the queue?

Reply:
I think the first test case `add highPriorityPodInfo to activeQ` already does this

Reply:
No, the first test case is passing (nil, pod) to Update(), that is just adding new Pod.
But, here it's passing (pod, pod) which is that pod doesn't exist in the queue. It means that "Update() just adds pod even when old pod doesn't exist", which shouldn't happen though.